docs/modules/record-encryption/con-record-encryption-overview.adoc

```diff
@@ -0,0 +1,95 @@
+
+Envelope encryption::
+Envelope encryption is an industry-standard technique suited for encrypting large volumes of data in an efficient manner.
+Data is encrypted with a Data Encryption Key (DEK).
+The DEK is encrypted using a Key Encryption Key (KEK).
+The KEK is stored securely in a Key Management System (KMS).
+Symmetric encryption keys::
+AES(GCM) 256 bit encryption symmetric encryption keys are used to encrypt and decrypt record data. 
+
+The process is as follows:
+
+. The filter intercepts produce requests from producing applications and encrypts the records.
+. The produce request is forwarded to the broker. 
+. The filter intercepts fetch responses from consuming applications and decrypts the records.
+. The fetch response is forwarded to the consuming application.
+
+The filter encrypts the record value only.
+Record keys, headers, and timestamps are not encrypted.
+
+The entire process is transparent from the point of view of Kafka clients and Kafka brokers. 
+Neither are aware that the records are being encrypted, nor do they have any access to the encryption keys or have any influence on the ciphering process to secure the records.
+
+== How the filter encrypts records
+The filter encrypts records from produce requests as follows:
+
+. Filter selects a KEK to apply.
+. Requests the KMS to generate a DEK for the KEK.
+. Uses an encrypted DEK (DEK encrypted with the KEK) to encrypt the record.
+. Replaces the original record with a cipher record (encrypted record, encrypted DEK, and metadata).
+
+The filter uses a DEK reuse strategy.
+Encrypted records are sent to the same topic using the same DEK until a time-out or an encryption limit is reached.
+
+== How the filter decrypts records
+The filter decrypts records from fetch responses as follows:
+
+. Filter receives a cipher record from the Kafka broker.
+. Reverses the process that constructed the cipher record.
+. Uses KMS to decrypt the DEK.
+. Uses the decrypted DEK to decrypt the encrypted record.
+. Replaces the cipher record with a decrypted record.
+
+The filter uses an LRU (least recently used) strategy for decrypted records.
+Decrypted DEKs are kept in memory to reduce interactions with the KMS.
+
+== How the filter uses the KMS
+To support the filter, the KMS provides the following:
+
+* A secure repository for storing Key Encryption Keys (KEKs)
+* A service for generating and decrypting Data Encryption Keys (DEKs)
+
+KEKs stay within the KMS.
+The KMS generates a DEK (which is securely generated random data) for a given KEK, then returns the DEK and an encrypted DEK.
+The encrypted DEK has the same data but encrypted with the KEK.
+The KMS doesn't store DEKs; they are stored as part of the cipher record in the broker.
+
+WARNING: The KMS must be available during runtime.
+If the KMS is unavailable, production and consumption through the filter become impossible until KMS service is restored.
```

Comment:
The DEK caches mitigate this somewhat. Though I'm not sure if we want to call that out.